substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrSymbolRepository.java

```diff
@@ -77,8 +76,6 @@ public long getSymbolId(String imageHeapString, boolean previousEpoch, boolean r
             return 0;
         }
 
-        assert Heap.getHeap().isInImageHeap(imageHeapString);
```

Comment:
This assertion was failing due to `com.sun.proxy.jdk.proxy1`. I think this assertion is incorrect because its possible we may sometimes need to serialize symbols not in the image heap.  This was exposed now since the ObjectCount event causes many more symbols to be required. What do you think?

Reply:
That would be bad, then we would need to change `JfrSymbolRepository` significantly. Do you have a stack trace?

Reply:
If the string is not in the image heap, would it be possible to derive the hash from the symbol string itself rather than from its memory address? 

I added an assertion to `JfrTypeRepository.writePackage`. Then with `mx native-unittest`:
```
21) com.oracle.svm.test.jfr.TestVirtualThreadsJfrStreaming#test
java.lang.AssertionError: Not in image heapcom.sun.proxy.jdk.proxy1
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.JfrTypeRepository.writePackage(JfrTypeRepository.java:179)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.JfrTypeRepository.writePackages(JfrTypeRepository.java:172)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.JfrTypeRepository.write(JfrTypeRepository.java:68)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.JfrChunkWriter.writeConstantPools(JfrChunkWriter.java:354)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.JfrChunkWriter.writeCheckpointEvent(JfrChunkWriter.java:324)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.JfrChunkWriter.writeFlushCheckpoint(JfrChunkWriter.java:296)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.JfrChunkWriter.closeFile(JfrChunkWriter.java:222)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jfr.SubstrateJVM.setOutput(SubstrateJVM.java:363)
        at jdk.jfr@21/jdk.jfr.internal.JVM.setOutput(JVM.java:167)
        at jdk.jfr@21/jdk.jfr.internal.MetadataRepository.setOutput(MetadataRepository.java:295)
        at jdk.jfr@21/jdk.jfr.internal.PlatformRecorder.start(PlatformRecorder.java:267)
        at jdk.jfr@21/jdk.jfr.internal.PlatformRecording.start(PlatformRecording.java:123)
        at jdk.jfr@21/jdk.jfr.consumer.RecordingStream.startAsync(RecordingStream.java:379)
        at com.oracle.svm.test.jfr.JfrStreamingTest.startStream(JfrStreamingTest.java:87)
        at com.oracle.svm.test.jfr.JfrStreamingTest.startStream(JfrStreamingTest.java:73)
        at com.oracle.svm.test.jfr.TestVirtualThreadsJfrStreaming.test(TestVirtualThreadsJfrStreaming.java:68)
        at java.base@21/java.lang.reflect.Method.invoke(Method.java:580)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
        at com.oracle.mxtool.junit.MxJUnitWrapper.runRequest(MxJUnitWrapper.java:376)
        at org.graalvm.nativeimage.junitsupport/com.oracle.svm.junit.SVMJUnitRunner.run(SVMJUnitRunner.java:164)
        at org.graalvm.nativeimage.junitsupport/com.oracle.svm.junit.SVMJUnitRunner.main(SVMJUnitRunner.java:184)
```

Reply:
>  please try to figure out where the problematic Package object is created

It gets created the first time we try to get the package. In this case, that's in `JfrTypeRepository` while we're trying to write out the type data. 
`Class.getPackage()` -->  `BootLoader.definePackage(Class<?> c)` --> `Target_jdk_internal_loader_BootLoader.getDefinedPackage` --> `Target_java_lang_Package()`
       
> I suppose that Package object also doesn't live in the image heap?

That's correct. It seems like when `DynamicHubInitializer.registerPackage` is called at build time, `javaClass.getPackage()` returns null for all packages with name `com.sun.proxy.jdk.proxy1`, preventing the package from being registered. Maybe this is why the package is not in the image heap.      

Is something wrong with the behavior for handling proxy classes? Otherwise, should we modify the `JfrSymbolRepository` (I can do that in a separate PR)?

Reply:
This sounds like a difference in behavior between Native Image and HotSpot. If HotSpot directly returns `null` for the package of proxy classes, then Native Image should do the same (i.e., it should not allocate a package at run-time in `Target_jdk_internal_loader_BootLoader.getDefinedPackage`).

Please try to implement a smaller reproducer that shows this difference in behavior and open a separate issue for this problem.

Reply:
Hi @christianhaeubl, I've opened an issue here: https://github.com/oracle/graal/issues/8796
It includes a small reproducer too.